src/antpm-usbmon2ant.1
.\" -*- coding: utf-8-unix -*-
.\" groff -man -Tutf8 antpm-downloader.1
.\" Copyright © 2013 by Kristóf Ralovich
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\"
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\" License.
.\"
.TH ANTPM-USBMON2ANT "1" "October 2013" "antpm-usbmon2ant v1.9" "User Commands"
.SH NAME
antpm-usbmon2ant \- Tool to convert ANT USB traffic traces
.SH DESCRIPTION
.SS "Allowed options:"
.TP
\fB\-h\fR [ \fB\-\-help\fR ]
produce help message
.TP
\fB\-o\fR [ \fB\-\-op\fR ] arg (=parse) possible modes of operation:
parse|dump|usbmon|filter|count
.TP
\fB\-P\fR [ \fB\-\-input\-file\fR ]
input file, if not given reads from standard input
.TP
\fB\-V\fR [ \fB\-\-version\fR ]
Print version information
.SH "AUTHOR"
.B antpm-downloader
was written by Kristóf Ralovich. This man page is Copyright ©
2010-2013 Kristóf Ralovich.
.SH "SEE ALSO"
.UR "https://github.com/ralovich/antpm"
.BR "https://github.com/ralovich/antpm"
.UE


debian/changelog
antpm (1.8-1) UNRELEASED; urgency=low

  * GarminConvert: add a method that offsets the timestamps
  * AntMessage: bugfix: correctly convert file creation time

 -- RALOVICH, Kristóf <tade60@freemail.hu>  Sat, 06 Jul 2013 00:46:18 +0200

antpm (1.7-1) UNRELEASED; urgency=low

  * antpm-*: recongize -h option for --help too
  * fit: this is a run-time warning
  * docs: start writing user's guide
  * docs: how to submit a bug report
  * docs: add a note on installing the udev rule for user access to the
    USB2ANT stick
  * docs: refine udev procedure
  * docs: split maintainer documentation from end-user documentation
  * FIT: bugfix: don't offset timestamps twice
  * bugfix: this was a crippling bug
  * docs: add todo
  * new version

 -- RALOVICH, Kristóf <tade60@freemail.hu>  Sat, 06 Jul 2013 00:11:34 +0200

antpm (1.6-3) UNRELEASED; urgency=low

  [ RALOVICH, Kristóf ]
  * docs: more notes
  * this change was only necessary for win64
  * debian: refine c++ compiler dependency
  * serial: add more verbosity to output

  [ Ralovich Kristof ]

 -- Ralovich Kristof <tade@cg.iit.bme.hu>  Wed, 26 Jun 2013 11:43:35 +0200

antpm (1.6-2) UNRELEASED; urgency=low

  [ Kristof Ralovich ]
  * script to generate changelog from git

  [ Ralovich Kristof ]

 -- Ralovich Kristof <tade@cg.iit.bme.hu>  Tue, 25 Jun 2013 17:10:28 +0200

antpm (1.6-0) UNRELEASED; urgency=low

  [ RALOVICH, Kristóf ]
  * include Release builds in a .deb. package
  * clarify build- and run-time dependencies
  * add todo
  * fix build with clang 3.0
  * add some documentation
  * fix build warning
  * git should ignore build folder
  * AntFr310XT2::createDownloadFolder()
  * further define folder name based on device S/N
  * cleanup
  * deb: lower build time dependencies
  * create download folder after knowing the device id
  * fit2gpx: option to decode ANT root folder
  * initial, file based logging
  * docs: note on performance
  * fix output
  * use centralized logging in more places
  * fix output
  * introduce namespace antpm
  * namespace conversion
  * missing namespace
  * more unified logging
  * more unified logging
  * cleanup
  * fix an assertion failure
  * unified logging replaced loggerc()
  * reindent
  * define NDEBUG in release
  * avoid crashes

  [ unknown ]
  * build fixes

  [ RALOVICH, Kristóf ]
  * add todo
  * add todo
  * move GARMIN_EPOCH to antdefs.hpp
  * bugfix: fix thinko
  * try setting date of .fit files
  * change logging
  * first stub for centralized per-device setting

  [ RALOVICH, Kristof ]
  * missing include
  * cmake: fix linking under win
  * previously the x86 dll hada weird name...
  * cmake: make sure the correct generator is invoked
  * fix 64-bit linking
  * cmake: fix determination of win64
  * scripts: build helper for win64
  * antpm-downloader: bugfix
  * update discriution list

  [ Kristof Ralovich ]
  * deb: relax build time dependencies

  [ RALOVICH, Kristof ]
  * deb: libusb was missing as a build-dep

  [ Kristof Ralovich ]
  * bugfix: resolve gcc-4.4 build issues

  [ RALOVICH, Kristof ]
  * pkgconfig should look for libusb-1.0

  [ Kristof Ralovich ]
  * cleanup
  * fix missing include
  * debian: fixing dependencies
  * account for boost:filesystem version changes

  [ RALOVICH, Kristof ]
  * cmake: fix finding libusb under nix

  [ Ralovich Kristof ]

 -- Ralovich Kristof <tade@cg.iit.bme.hu>  Tue, 25 Jun 2013 17:02:17 +0200

antpm (1.0-1) UNRELEASED; urgency=low

  * Initial release.

 -- Kristof Ralovich <tade60@freemail.hu>  Tue, 21 Mar 2013 16:22:10 +0100


src/antpm-usbmon2ant.cpp
// -*- mode: c++; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2; coding: utf-8-unix -*-
// ***** BEGIN LICENSE BLOCK *****
//////////////////////////////////////////////////////////////////////////
// Copyright (c) 2011-2014 RALOVICH, Kristóf                            //
//                                                                      //
// This program is free software; you can redistribute it and/or modify //
// it under the terms of the GNU General Public License as published by //
// the Free Software Foundation; either version 3 of the License, or    //
// (at your option) any later version.                                  //
//                                                                      //
// This program is distributed in the hope that it will be useful,      //
// but WITHOUT ANY WARRANTY; without even the implied warranty of       //
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the        //
// GNU General Public License for more details.                         //
//                                                                      //
//////////////////////////////////////////////////////////////////////////
// ***** END LICENSE BLOCK *****

#include <cstdio>
#include <cstdlib>
#include <iostream>
#include <fstream>
#include <vector>

#include "AntMessage.hpp"
#include "common.hpp"
#include "Log.hpp"
#include "GarminPacketIntf.hpp"
#include "garmintools/garmin.h"

#include <boost/program_options.hpp>
#include <boost/exception/diagnostic_information.hpp>

namespace po = boost::program_options;
using namespace std;
using namespace antpm;

DEFAULT_LOG_INSTANTIATOR

namespace antpm
{

// Interpret a sequence of burst messages while replaying a trace file.
class BurstDecodeVisitor
{
public:
  BurstDecodeVisitor() { mFilterDirect=false; }
  void onMessage(const AntMessage& m);
  void onLastBurst(const uchar chan);
  bool tryDecodeDirect(const uchar chan, std::vector<uint8_t> &burstData);
  bool tryFilterDirect(const uchar chan, std::vector<uint8_t> &burstData);
  void printDirect(const char *dir, std::vector<uint8_t> &burstData);
protected:
  map<uchar, list<AntMessage> > chanBursts; // <channel, aggregated burst> pairs
  map<uchar, int> chanLastCmd;              // <channel, pid> pairs
  bool mFilterDirect;
};


void
BurstDecodeVisitor::onMessage(const AntMessage &m)
{
  if(m.bytes.size()<3 || m.getMsgId()!=MESG_BURST_DATA_ID || m.getLenPayload()!=9)
    return;
  const M_ANT_Burst* burst(reinterpret_cast<const M_ANT_Burst*>(m.getPayloadRef()));

  uchar chan = burst->chan;


  chanBursts[chan].push_back(m);

  if(burst->isLast())
    onLastBurst(chan);
}

void
BurstDecodeVisitor::onLastBurst(const uchar chan)
{
  list<AntMessage>& msgs = chanBursts[chan];
  std::vector<uint8_t> burstData;
  uchar expectedSeq=0;
  //bool found=false;
  bool lastFound=false;

  for(std::list<AntMessage>::iterator i = msgs.begin(); i != msgs.end(); i++)
  {
    AntMessage& repl(*i);
    assert(repl.getLenPayload()==9);
    const M_ANT_Burst* burst(reinterpret_cast<const M_ANT_Burst*>(repl.getPayloadRef()));
    //CHECK_RETURN_FALSE(burst->seq == expectedSeq);
    ++expectedSeq;
    if(expectedSeq == 4)
      expectedSeq = 1;
    //found = true;
    std::vector<uchar> crtBurst(repl.getPayload());
    burstData.insert(burstData.end(), crtBurst.begin()+1,crtBurst.end());
    assert(burstData.size()%8==0);
    lastFound = burst->isLast();
    if(lastFound) break;
  }
  msgs.clear();

  if(!mFilterDirect)
    tryDecodeDirect(chan, burstData);
  else
    tryFilterDirect(chan, burstData);
}


bool
BurstDecodeVisitor::tryDecodeDirect(const uchar chan, std::vector<uint8_t>& burstData)
{
  CHECK_RETURN_FALSE(burstData.size()>=8);
  const M_ANTFS_Beacon* beac(reinterpret_cast<const M_ANTFS_Beacon*>(&burstData[0]));
  const M_ANTFS_Command* cmd(reinterpret_cast<const M_ANTFS_Command*>(&burstData[0]));
  std::vector<uint8_t> data;
  GarminPacketIntf gpi;

  if(beac->beaconId==ANTFS_BeaconId)
  {
    // beacon+response
    CHECK_RETURN_FALSE(burstData.size()>=2*8);
    const M_ANTFS_Response* resp(reinterpret_cast<const M_ANTFS_Response*>(&burstData[8]));
    CHECK_RETURN_FALSE(resp->responseId==ANTFS_CommandResponseId);
    //CHECK_RETURN_FALSE(resp->response==ANTFS_RespDirect);
    if(resp->response!=ANTFS_RespDirect) return false;

    logger() << "expecting " << resp->detail.directResponse.data << "x8 bytes of direct data, plus 16 bytes\n";
    logger() << "got back = \"" << burstData.size() << "\" bytes\n";
    CHECK_RETURN_FALSE(burstData.size()==size_t((2+resp->detail.directResponse.data)*8));

    data = vector<uint8_t>(burstData.begin()+16, burstData.end());

    bool rv = gpi.interpret(this->chanLastCmd[chan], data);
    this->chanLastCmd[chan] = -1;
    return rv;
  }
  else if(cmd->commandId==ANTFS_CommandResponseId)
  {
    // command
    //CHECK_RETURN_FALSE(cmd->command==ANTFS_CmdDirect);
    if(cmd->command!=ANTFS_CmdDirect) return false;

    logger() << "got = \"" << burstData.size() << "\" bytes\n";
    CHECK_RETURN_FALSE(burstData.size()==size_t((1+1)*8));

    data = vector<uint8_t>(burstData.begin()+8, burstData.end());

    // save the last command, we expect to decode the corresponding response based on this
    this->chanLastCmd[chan] = gpi.interpretPid(data);

    return gpi.interpret(-1, data);
  }
  else
  {
    LOG(LOG_WARN) << "Couldn't decode burst data starting with 0x" << toString<int>((int)burstData[0], 2, '0') << " !\n";
    return false;
  }
}


bool
BurstDecodeVisitor::tryFilterDirect(const uchar chan, std::vector<uint8_t> &burstData)
{
  CHECK_RETURN_FALSE(burstData.size()>=8);
  const M_ANTFS_Beacon* beac(reinterpret_cast<const M_ANTFS_Beacon*>(&burstData[0]));
  const M_ANTFS_Command* cmd(reinterpret_cast<const M_ANTFS_Command*>(&burstData[0]));
  std::vector<uint8_t> data;
  //GarminPacketIntf gpi;

  if(beac->beaconId==ANTFS_BeaconId)
  {
    // beacon+response
    CHECK_RETURN_FALSE(burstData.size()>=2*8);
    const M_ANTFS_Response* resp(reinterpret_cast<const M_ANTFS_Response*>(&burstData[8]));
    CHECK_RETURN_FALSE(resp->responseId==ANTFS_CommandResponseId);
    //CHECK_RETURN_FALSE(resp->response==ANTFS_RespDirect);
    if(resp->response!=ANTFS_RespDirect) return false;

    logger() << "expecting " << resp->detail.directResponse.data << "x8 bytes of direct data, plus 16 bytes\n";
    logger() << "got back = \"" << burstData.size() << "\" bytes\n";
    CHECK_RETURN_FALSE(burstData.size()==size_t((2+resp->detail.directResponse.data)*8));

    data = vector<uint8_t>(burstData.begin()+16, burstData.end());

    this->printDirect("R ", data);

    return true;
  }
  else if(cmd->commandId==ANTFS_CommandResponseId)
  {
    // command
    //CHECK_RETURN_FALSE(cmd->command==ANTFS_CmdDirect);
    if(cmd->command!=ANTFS_CmdDirect) return false;

    logger() << "got = \"" << burstData.size() << "\" bytes\n";
    CHECK_RETURN_FALSE(burstData.size()==size_t((1+1)*8));

    data = vector<uint8_t>(burstData.begin()+8, burstData.end());

    this->printDirect("S ", data);

    return true;
  }
  else
  {
    LOG(LOG_WARN) << "Couldn't decode burst data starting with 0x" << toString<int>((int)burstData[0], 2, '0') << " !\n";
    return false;
  }

}

void BurstDecodeVisitor::printDirect(const char* dir, std::vector<uint8_t> &data)
{
  cout << "@#@#: " << dir;
  for(size_t i = 0; i < data.size(); i++)
  {
    cout << toString<int>(data[i], 2, '0');
  }
  cout << " ";
  for(size_t i = 0; i < data.size(); i++)
  {
    cout << char(isprint(data[i])?data[i]:'.');
  }
  cout << "\n";
}

}




int
main(int argc, char** argv)
{
#ifndef NDEBUG
  antpm::Log::instance()->setLogReportingLevel(antpm::LOG_DBG);
#else
  antpm::Log::instance()->setLogReportingLevel(antpm::LOG_INF);
#endif

  istream* in=0;
  ifstream f;

  // Declare the supported options.
  //bool dump=false;
  //bool usbmon=false;
  //bool filter=false;
  string op;
  string inputFile;
  po::positional_options_description pd;
  pd.add("input-file", 1);
  po::options_description desc("Allowed options");
  po::variables_map vm;
  try
  {
    desc.add_options()
    ("help,h",                                                       "produce help message")
    ("op,O", po::value<string>(&op)->default_value("parse"), "possible modes of operation: parse|dump|usbmon|filter|count")
    //("d", po::value<bool>(&dump)->zero_tokens(), "diffable byte dumps + decoded strings")
    //("u", po::value<bool>(&usbmon)->zero_tokens(), "generate pseudo usbmon output")
    //("f", po::value<bool>(&filter)->zero_tokens(), "just filter ANT messages from usbmon stream")
    ("input-file,P", po::value<string>(&inputFile)->zero_tokens(),   "input file, if not given reads from standard input")
    ("version,V",                                               "Print version information")
    ;

    //po::parsed_options parsed = po::parse_command_line(argc, argv, desc);
    po::parsed_options parsed = po::command_line_parser(argc, argv).options(desc).positional(pd).run();
    po::store(parsed, vm);
    po::notify(vm);
  }
  catch(po::error& error)
  {
    cerr << error.what() << "\n";
    cerr << desc << "\n";
    return EXIT_FAILURE;
  }
  catch(boost::exception& e)
  {
    cerr << boost::diagnostic_information(e) << std::endl;
    return EXIT_FAILURE;
  }
  catch(std::exception& ex)
  {
    cerr << ex.what() << "\n";
    cerr << desc << "\n";
    return EXIT_FAILURE;
  }


  if(vm.count("version") || vm.count("V"))
  {
    cout << argv[0] << " " << antpm::getVersionString() << "\n";
    return EXIT_SUCCESS;
  }

  if(vm.count("help"))
  {
    cout << desc << "\n";
    return EXIT_SUCCESS;
  }
  //cout << dump << "\n";
  //cout << inputFile << "\n";

  if(!inputFile.empty())
  {
    f.open(inputFile.c_str());
    if(!f.is_open())
      return EXIT_FAILURE;
    in=&f;
  }
  else
  {
    in=&cin;
  }

  //if(argc==2)
  //{
  //  f.open(argv[1]);
  //  if(!f.is_open())
  //    return EXIT_FAILURE;
  //  in=&f;
  //}
  //else
  //{
  //  in=&cin;
  //}

  std::list<AntMessage> messages;
  BurstDecodeVisitor burstDecoder;

  size_t lineno=0;
  std::string line;
  unsigned long timeUs0=0;
  bool first=true;
  for( ;getline(*in, line); lineno++)
  {
    vector<string> vs(split(line, ' '));
    //cout << vs.size() << "\n";

    if(vs.size()<8 || vs[6]!="=" || vs[7].size()<2 || vs[7][0]!='a' || vs[7][1]!='4')
      continue;
    
    unsigned long timeUs=0;
    if(1!=sscanf(vs[1].c_str(), "%lu", &timeUs))
      continue;
    if(first)
    {
      first=false;
      timeUs0=timeUs;
    }
    double dt=(timeUs-timeUs0)/1000.0;
    timeUs0=timeUs;

    unsigned long usbExpectedBytes=0;
    if(1!=sscanf(vs[5].c_str(), "%lu", &usbExpectedBytes))
      continue;

    //cout << ".\n";
    std::string buf;
    for(size_t i=7;i<vs.size();i++)
      buf+=vs[i];
    //cout << buf << "\n";

    std::list<uchar> bytes;
    std::vector<AntMessage> messages2;

    if(!AntMessage::stringToBytes(buf.c_str(), bytes))
    {
      cerr << "\nDECODE FAILED [1]: line: " << lineno << " \"" << line << "\"!\n\n";
      continue;
    }
    size_t usbActualBytes=bytes.size();
    if(!AntMessage::interpret2(bytes, messages2))
    {
      if(usbActualBytes<usbExpectedBytes && !messages2.empty())
      {
        // long line truncated in usbmon, but we still managed to decoded one or more packets from it
        cerr << "\nTRUNCATED: line: " << lineno << " \"" << line << "\" decoded=" << messages2.size() << "!\n\n";
      }
      else
      {
        cerr << "\nDECODE FAILED [2]: line: " << lineno << " \"" << line << "\"!\n\n";
        continue;
      }
    }

    if(op=="count")
    {
      cout << lineno << "\t" << messages2.size() << "\n";
      continue;
    }
    else if(op=="filter")
    {
      cout << line << "\n";
      continue;
    }

    for (std::vector<AntMessage>::iterator i=messages2.begin(); i!=messages2.end(); i++)
    {
      i->idx = lineno;
      i->sent = (vs[2][0]=='S');
      AntMessage& m=*i;
      
      if(op=="usbmon")
      {
        messages.push_back(m);
      }
      else if(op=="dump")
      {
        cout << m.strExt() << "\n";
      }
      else
      {
        //cout << dt << "\t" << m.str() << "\n";
        cout << m.strDt(dt) << "\n";
        burstDecoder.onMessage(m);
      }
    }
  }

  if(op=="usbmon")
  {
    AntMessage::saveAsUsbMon(cout, messages);
  }

  return EXIT_SUCCESS;
}


